spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/ComputePartitionStatsSparkAction.java

```diff
@@ -0,0 +1,102 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.actions;
+
+import java.io.IOException;
+import org.apache.iceberg.PartitionStatisticsFile;
+import org.apache.iceberg.PartitionStatsHandler;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.actions.ComputePartitionStats;
+import org.apache.iceberg.actions.ImmutableComputePartitionStats;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.spark.JobGroupInfo;
+import org.apache.spark.sql.SparkSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Computes the stats incrementally after the snapshot that has partition stats file till the given
+ * snapshot (uses current snapshot if not specified) and writes the combined result into a {@link
+ * PartitionStatisticsFile} after merging the stats for a given snapshot. Does a full compute if
+ * previous statistics file does not exist. Also registers the {@link PartitionStatisticsFile} to
+ * table metadata.
+ */
+public class ComputePartitionStatsSparkAction
+    extends BaseSparkAction<ComputePartitionStatsSparkAction> implements ComputePartitionStats {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ComputePartitionStatsSparkAction.class);
+  private static final Result EMPTY_RESULT =
+      ImmutableComputePartitionStats.Result.builder().build();
+
+  private final Table table;
+  private Snapshot snapshot;
+
+  ComputePartitionStatsSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    this.snapshot = table.currentSnapshot();
+  }
+
+  @Override
+  protected ComputePartitionStatsSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public ComputePartitionStats snapshot(long newSnapshotId) {
+    Snapshot newSnapshot = table.snapshot(newSnapshotId);
+    Preconditions.checkArgument(newSnapshot != null, "Snapshot not found: %s", newSnapshotId);
+    this.snapshot = newSnapshot;
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    if (snapshot == null) {
+      LOG.info("No snapshot to compute partition stats for table {}", table.name());
+      return EMPTY_RESULT;
+    }
+
+    JobGroupInfo info = newJobGroupInfo("COMPUTE-PARTITION-STATS", jobDesc());
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    LOG.info("Computing partition stats for {} (snapshot {})", table.name(), snapshot.snapshotId());
+    PartitionStatisticsFile statisticsFile;
+    try {
+      statisticsFile = PartitionStatsHandler.computeAndWriteStatsFile(table, snapshot.snapshotId());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    if (statisticsFile == null) {
+      return EMPTY_RESULT;
+    }
+
+    table.updatePartitionStatistics().setPartitionStatistics(statisticsFile).commit();
```

Comment:
If the statistics for a particular snapshot were already computed, and user triggered the action second time what is the behavior?
1. Does the action recompute? Or does it return the existing statistics file
2. If it returned the existing file from first computation, how does the user know if the stats in result is from a prior computation. Do we need a field in result to denote this, like `CREATED` (when computed fresh), `EXISTING` (when it returns an already computed one)?
3. Related to 2, if we are returning an existing stats file, do we need a way to force the computation?

Reply:
> If the statistics for a particular snapshot were already computed, and user triggered the action second time what is the behavior?

Currently, it reads the stats file  (of the same snapshot) and write it into a new stats file without any additional compute of reading manifests. I will change it to no-op by returning the same file along with adding Info logs that stats file exist for snapshot. 

> do we need a way to force the computation

The user can unregister the stats from the table using `table.updatePartitionStatistics().removePartitionStatistics(snapshotId).commit();`, so it can force recompute when the compute API is called again.

Reply:
> change it to no-op

+1
I will let you take a call on whether we need to return whether it was created or an existing file or do it in a subsequent PR .

For a sql user, there is no way to figure why the command completes faster(when returning existing) vs taking time(when computed new) without a status.

Although this might not be an issue using APIs, since the user can check whether stats file is already available using table api (`org.apache.iceberg.Table#partitionStatisticsFiles`).

Reply:
> For a sql user, there is no way to figure why the command completes faster(when returning existing) vs taking time(when computed new) without a status.

Maybe we need to provide a metadata table to list the statistics file (both table and partition) in future. 

>  we need to return whether it was created or an existing file

I think adding status is an overkill. They think about freshness etc. I will make it no op and print in the logs about why it is no op.

Reply:
Added no op and logs + testcase.